Refactor fetch_url and extend tests

nickmoreton opened the pull request:
# Implement the Responses Library

I have worked through any tests that needed to work with external site urls and implemented the responses library in those place. I've also add some extra tests that I felt needed to be done.

I made changes to the CaptionHandler().construct_block method so the tests are easier to implement as well as using responses library.

**There is one TODO** I added in the above method to implement logging for a case of a caption shortcode that has no `<img` tag. There's an [upcoming ticket](https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/63) to improve the logging. It's not a part of the process that will prevent an import continuing and adds a raw_html block with a helpful message thats not visible in the frontend by using HTML <!-- comment -->.

Ticket URL: https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/76

---

<!-- Please tick or remove these as relevant. Provide further details if valuable. Be pragmatic. -->

- Testing
    - [x] CI passes
    - [x] If necessary, tests are added for new or fixed behaviour
    - [x] These changes do not reduce test coverage
- Documentation.
    - [x] Documentation changes are not necessary because: there are no changes to the usage of the package.

nickmoreton committed:
refactor fetch_url and exptend tests

save

lint

nickmoreton committed:
update ci dependencies

save

save

save

save

save

nickmoreton committed:
change fixture to remove live domain

update tests to mock images and pdf content

nickmoreton committed:
revert earlier chnages

nickmoreton committed:
add new install dependencies

nickmoreton committed:
try adding dependencies as testing only

save

save

save

nickmoreton committed:
replace domain name with example.com

nickmoreton committed:
add extra test for absence of wp:postmeta in caches

nickmoreton committed:
add test utility functions module

lint

nickmoreton committed:
add image mock to tests for shortcode handlers

I changed construct_block of the caption handler to avoid using so many try except blocks and we can easily test the outputs.

nickmoreton committed:
get the actuall image id by searching the image by title

nimasmi commented on wagtail_wordpress_import/block_builder_defaults.py:
```diff
@@ -238,16 +238,16 @@ def get_or_save_image(src):
-        r = requests.get(src, **conf_get_requests_settings())
-        status = r.status_code == 200
+        response = requests.get(src, **conf_get_requests_settings())
+        status = True if response.status_code == 200 else False
```
**Question**: What is this change for? Just readability?

nickmoreton replied:
Yes I felt the return names needed to be clearer too.

nimasmi commented on wagtail_wordpress_import/block_builder_defaults.py:
```diff
@@ -238,16 +238,16 @@ def get_or_save_image(src):
-    return r, status, content_type
+        return response, status, content_type
+    except Exception as e:
+        raise requests.ConnectionError(e)
```
**Question**: Why catch anything here if we're not going to handle it? Are the other types of exception not `ConnectionError`?

nickmoreton replied:
There seems to be a few possible exception types from requests and to handle all the possible values might be overkill but we do need to avoid the import process from stopping for any exception.

We might change this when we do the ticket for improving the logging (there's no logging for this here other than in the console).

nimasmi replied:
Okay, so why catch and reraise? Is it because you want to raise a single error type that you can watch for later?

nimasmi replied:
@nickmoreton Can you please address this comment?

nickmoreton replied:
> Okay, so why catch and reraise? Is it because you want to raise a single error type that you can watch for later?

Yes I was thinking in the later logging improvements we'd log a specific error

nimasmi commented on wagtail_wordpress_import/test/tests/utility_functions.py:
```diff
@@ -0,0 +1,21 @@
+    temp_file = tempfile.NamedTemporaryFile(suffix=".jpg")
+    image = Image.new("RGB", (200, 200), "white")
+    image.save(temp_file, "PNG")
+    return open(temp_file.name, mode="rb")
```
**Issue**: We are using the suffix ".jpg" for a PNG image. It's probably harmless since we're not expecting to process it, but it could lead to confusion in future.

**Question**: Was saving these as tempfiles necessary? Would a `io.BytesIO` object work?

nimasmi replied:
The issue is fixed in 98c1cfe.

nimasmi replied:
The image tempfile issue is fixed in 2a2278d.

I meant that this is an issue for all functions using tempfile, but probably it can be left if you prefer. In general the use of the filesystem is nice to avoid if possible during unit tests, unless you _need_ to test that something is correctly read or written.

nimasmi replied:
@nickmoreton Please respond, or resolve this thread, as necessary.

nimasmi commented on wagtail_wordpress_import/test/tests/test_import_hooks_items.py:
```diff
@@ -124,6 +124,11 @@ def test_xml_items_are_cached_without_duplicates(self):
+
+        # there should be no wp:postmeta in the cache
+        with self.assertRaises(KeyError):
+            foo["wp:postmeta"]
```
**Question**: I'm not sure of the relevance of this change to the URL refactoring. What's the link?

nickmoreton replied:
I noticed that we had no test for that when I was running coverage and saw it as an opportunity to improve coverage.

nimasmi replied:
@nickmoreton Please respond to this question.

nickmoreton replied:
This is still pending too.

nimasmi replied:
Okay, cheers. It makes sense to add it.

nimasmi commented on wagtail_wordpress_import/test/tests/test_block_builder.py:
```diff
@@ -284,9 +353,32 @@ def test_get_alignment_class_not_present(self):
+            "https://www.example.com/images/connection_error.jpg",
+            body=Exception("Connection error"),
+        )
+        with self.assertRaises(Exception) as ctx:
```
**Question**: Are we testing for `requests.ConnectionError` here? If so, can we assert that is raised instead.?

nimasmi replied:
This is fixed in 75b1ad8.

nimasmi commented on wagtail_wordpress_import/test/tests/test_block_builder.py:
```diff
@@ -201,32 +219,83 @@ def test_richtext_block_6_content(self):
+        self.assertEqual(embed_tag.attrs["id"], str(image.id))
+        self.assertEqual(embed_tag.attrs["embedtype"], "image")
+        self.assertEqual(embed_tag.attrs["format"], "fullwidth")
+        self.assertEqual(embed_tag.attrs["alt"], "bruno 4 runner")
```
**Question**: What are we testing here? I just want to understand this.

- 238 we build the HTML with an embedded image, as though it is being imported (which fetches the image URL)
- 239 we assume the image is stored, and retrieve it from the database
- 242–245 we check that the formatted HTML is correct, and matches the expected image

?

nimasmi replied:
**Suggestion**: As an alternative to inline comments, you can use `TestCase.subTest` to add comments, and this also gives context in the output if the test fails during the subTest block.

```suggestion
        with self.subTest("the image tag is processed successfully"):
            output = image_linker(raw_html_file)
        with self.subTest("the image is stored in the database"):
            image = get_image_model().objects.get(title="bruno-4-runner.jpg")
        soup = BeautifulSoup(output, "html.parser")
        with self.subTest("the HTML is formatted with the image properties"):
            embed_tag = soup.find("embed")
            self.assertEqual(embed_tag.attrs["id"], str(image.id))
            self.assertEqual(embed_tag.attrs["embedtype"], "image")
            self.assertEqual(embed_tag.attrs["format"], "fullwidth")
            self.assertEqual(embed_tag.attrs["alt"], "bruno 4 runner")
```

nimasmi replied:
This is fixed in 3479fba.

nimasmi commented on wagtail_wordpress_import/test/tests/test_shortcode_handlers.py:
```diff
@@ -274,13 +275,34 @@ class FooHandler(BlockShortcodeHandler):
+
+        foo = FooHandler()
+        is_top_level_html_tag = foo.is_top_level_html_tag
+        self.assertEqual(is_top_level_html_tag, False)
```
**Question**: What are we trying to test here? This test calls no code outside of the test case method (other than instantiating a class that subclasses something from elsewhere, but still it only tests a property that is overwritten within this test case).

nimasmi replied:
@nickmoreton Please respond to this question.

nimasmi replied:
This is changed in ac1ecef^..9157859

nimasmi commented on wagtail_wordpress_import/test/tests/test_shortcode_handlers.py:
```diff
@@ -304,3 +326,136 @@ def test_construct_block_method_output(self):
+            <a href="http://www.example.com/">
+                
+            </a>This is a caption about the image
+        </wagtail_block_caption>"""
```
**Question**: We're testing that if the image is not found at the URL, there's no image in the imported document? Why add the response at all?

nimasmi replied:
This is resolved in 150afdd.

nimasmi commented on wagtail_wordpress_import/prefilters/handle_shortcodes.py:
```diff
@@ -134,36 +134,44 @@ def construct_block(self, soup):
+        image_file = get_or_save_image(image.attrs["src"])
+
+        # Parse alignment
+        if "align" in soup.attrs:
+            alignment = soup.attrs["align"].replace("align", "")
+        else:
+            alignment = "left"
```
**Question**: Why do we have different alignment parsing, one here expecting "alignright" and the other in `wagtail_wordpress_import.block_builder_defaults.get_alignment_class` expecting "alignright"? Is it because two different sources are formatted differently?

nickmoreton replied:
The difference is that we are using the align value to set the alignment in
```
class ImageBlock(blocks.StructBlock):
    image = ImageChooserBlock()
    caption = blocks.CharBlock(required=False)
    link = blocks.URLBlock(required=False)
    alignment = blocks.ChoiceBlock(
        choices=(
            ("left", "Left"),
            ("right", "Right"),
            ("center", "Center"),
        ),
        default="left",
    )
```
 for a caption shortcode and get_alignment_class() is used in the rich_text editor to set a css class

nimasmi replied:
@nickmoreton Can you respond to this question, please?

nickmoreton replied:
I responded here I think but it's showing as pending for me. Does that mean you can't see it?

nimasmi replied:
Okay, thanks. I can see your response now. That explanation is helpful.

nickmoreton committed:
change image file suffix

nickmoreton committed:
convert mock_image () to use io.BytesIO

also updated test to reflecet the change

nickmoreton committed:
assert test us for a specific error

nickmoreton committed:
use subTest for better error feedback

nickmoreton committed:
change class attribute name

The attribute name was the same as the class method is_top_level_html_tag()

I also added a default return type in is_top_level_html_tag() to avoid errors during the import porcess.

nickmoreton committed:
add test for the default value of top_level_html_tag

save

nickmoreton committed:
remove unnecessary response

nickmoreton committed:
remove earlier TODO comment

I also altered the test to check for the real image.id

save

nimasmi commented on wagtail_wordpress_import/prefilters/handle_shortcodes.py:
```diff
@@ -96,7 +96,9 @@ def element_name(self):
-        return self.is_top_level_html_tag
+        if not type(self.top_level_html_tag) == bool:
+            return True
+        return self.top_level_html_tag
```
**Comment**: I don't understand the way this was originally written. The following class:

```python
class MyClass:
    is_foo = True

    @property
    def is_foo(self):
        return self.is_foo
```

gives the following REPL session:

```python
>>> inst = MyClass()
>>> inst.is_foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nick/projects/wxd/wagtail-wordpress-import/thing.py", line 6, in is_foo
    return self.is_foo
  File "/home/nick/projects/wxd/wagtail-wordpress-import/thing.py", line 6, in is_foo
    return self.is_foo
  File "/home/nick/projects/wxd/wagtail-wordpress-import/thing.py", line 6, in is_foo
    return self.is_foo
  [Previous line repeated 996 more times]
RecursionError: maximum recursion depth exceeded
>>> 
```

Why do we need to test that this is a boolean type? Why do we need a method to access the attribute?

nimasmi replied:
This approach has been removed.

nimasmi commented on wagtail_wordpress_import/test/tests/test_shortcode_handlers.py:
```diff
@@ -274,13 +275,75 @@ class FooHandler(BlockShortcodeHandler):
+            top_level_html_tag = {"foo": "bar"}
+
+        foo = FooHandler()
+        self.assertTrue(foo.is_top_level_html_tag)
```
Discussion about this test case was originally started in https://github.com/torchbox/wagtail-wordpress-import/pull/100#discussion_r760992232 (now outdated).

Thanks for adding the docstring. I don't follow why we need to handle the case where this is set as a non-boolean. Some of those cases are very counterintuitive, e.g. why would setting `top_level_html_tag = None` give `obj.is_top_level_html_tag == True`? I would prefer it if setting a non-boolean would be the default truthiness of the object.

I would suggest removing the `is_...` method, and restoring it to a bare attribute (see https://github.com/torchbox/wagtail-wordpress-import/pull/100/files#r761904862), and testing only for:

```suggestion
class TestProvidedShortcodeIsTopLevel(TestCase):
    """
    Block Shortcode Handler has a top_level_property that is True by default.
    It can be set to False to indicate that the shortcode is not a top level HTML tag.
    """

    def test_is_top_level_html_tag_defaults_to_true(self):
        class FooHandler(BlockShortcodeHandler):
            pass

        foo = FooHandler()
        self.assertTrue(foo.is_top_level_html_tag)

    def test_is_top_level_html_tag_can_be_overridden(self):
        class FooHandler(BlockShortcodeHandler):
            is_top_level_html_tag = False

        foo = FooHandler()
        self.assertFalse(foo.is_top_level_html_tag)
```

nimasmi replied:
This is changed in ac1ecef..9157859

nimasmi commented on wagtail_wordpress_import/test/tests/test_shortcode_handlers.py:
```diff
@@ -291,16 +354,140 @@ def test_construct_block_method_output(self):
-        # there is a ticket to improve testing when fetching remote images:
-        # https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/76
-        self.assertIsInstance(int(json["value"]["image"]), int)
+        self.assertEqual(json["value"]["image"], image.id)
```
**Question**: Wasn't this also fixed in #98?

nickmoreton replied:
#98 was for the test_block_builder TODO

nimasmi replied:
This is answered in https://github.com/torchbox/wagtail-wordpress-import/pull/100#issuecomment-985507657

nickmoreton committed:
Revert "change class attribute name"

This reverts commit 6bea19952de4d50b6255b072c8beb6c1ac93037d.

nickmoreton committed:
Revert "add test for the default value of top_level_html_tag"

This reverts commit 95d0fc4641abb05f2f6930b24d93ca632c0d034d.

nickmoreton committed:
remove is_top_level_html_tag()

In fact the implementation of is_top_level_html_tag: True was wrong it shoul have been is_top_level_html_tag = True

nickmoreton committed:
chaneg tests to reflect the simpler way to do implement this

nickmoreton reviewed (commented):
I probably didn't finish the review. Hopefully this works now :)
